plutus-core/docs/BuiltinsOverview.md

```diff
@@ -0,0 +1,456 @@
+3. The "meaning" of each built-in function, more on that later.
+
+A `BuiltinCostModel` lives in the `Evaluation.Machine.ExBudgetingDefaults` module and all other costing ingridients can be discovered from there, but we're going to focus on non-costing parts of the built-in functions machinery.
+
+### Adding a built-in function
+
+Costing issues aside, the only module one needs to amend in order to define a new built-in function for UPLC/TPLC/PIR is `Default.Builtins` (for Plutus Tx one has also edit modules in `plutus-tx` and `plutus-tx-plugin`, but that is out of scope of this document). For that do the following:
+
+1. Add a new constructor to `DefaultFun`.
+2. Handle it in the `toBuiltinMeaning` method in the `ToBuiltinMeaning uni DefaultFun` instance. You'll get a warning there once you've added the builtin to `DefaultFun`.
+3. Handle the constructor in the `encode` and `decode` methods in the `Flat DefaultFun` instance. You'll only get a warning in the former once you've added the builtin to `DefaultFun` and not the latter unfortunately, so make sure not to forget about it.
+
+And that's it, you'll get everything else automatically, including general tests that are not specific to the particular builtin (please write specific tests for new builtins!). If you run the tests and then check `git status`, you'll find a new golden file that contains the Plutus type of the new builtin.
+
+`Default.Builtins` contains extensive documentation on how to add a built-in function, what is allowed and what should be avoided, make sure to read everything if you want to add a built-in function.
+
+In practice, adding new builtins is more complicated than that, because some of the tests and Plutus Tx definitions need to be amended/created manually, see [this](https://github.com/input-output-hk/plutus/commit/173dce5ee85cb8038563dd39299abb550ea13b88) commit for an example. This is on top of costing requiring a lot of manual labor.
+
+### Builtin meanings
+
+`toBuiltinMeaning` is the sole method of the `ToBuiltinMeaning` type class (not counting two associated type/data families) defined in `Builtin.Meaning` and is used for assigning each built-in function its `BuiltinMeaning`, which comprises everything needed for testing, type checking and executing the builtin. The type signature of the method looks like this:
+
+```haskell
+toBuiltinMeaning
+    :: HasMeaningIn uni val
+    => BuiltinVersion fun -> fun -> BuiltinMeaning val (CostingPart uni fun)
+```
+
+i.e. in order to construct a `BuiltinMeaning` one needs not only a built-in function, but also a version of the set of built-in functions. You can read more about versioning of builtins and everything else in [CIP-35](https://cips.cardano.org/cips/cip35).
```

Comment:
Are you referring to the recently renamed BuiltinSemanticsVariant?

Reply:
Yes, thank you, will fix.